proposals/compile-time-dependency-resolution.md

```diff
@@ -0,0 +1,489 @@
+
+```kotlin
+data class Group<A>(val values: List<A>)
+
+// Kotlin + KEEP-87
+extension class GroupRepository<A>(with val R: Repository<A>) : Repository<Group<A>> {
+  override fun loadAll(): Group<A> =
+    Group(R.loadAll())
+}
+
+// Regular Kotlin
+class GroupRepository<A>(val R: Repository<A>) : Repository<Group<A>> {
+  override fun loadAll(): Group<A> = with (R) {
+    Group(loadAll())
+  }
+}
+```
+
+## Composition and chain of evidences
+
+Constraint interface declarations and extension evidences can encode further constraints on their type parameters so that they can be composed nicely:
+
+```kotlin
+package com.data.instances
+
+import com.data.Repository
+import com.domain.User
+import com.domain.Group
+
+extension class GroupRepository<A>(with val repoA: Repository<A>) : Repository<Group<A>> {
+  override fun loadAll(): List<Group<A>> {
+    return listOf(Group(repoA.loadAll()))
+  }
+
+  override fun loadById(id: Int): Group<A>? {
+    return Group(repoA.loadById(id))
+  }
+}
+```
+
+The above extension provides evidence of a `Repository<Group<A>>` as long as there is a `Repository<A>` in scope. Call site would be like:
+
+```kotlin
+fun fetchGroup<A>(with repo: GroupRepository<A>) = repo.loadAll()
+
+fun main() {
+  fetchGroup<User>() // Succeeds! There's evidence of Repository<Group<A>> and Repository<User> provided in scope.
+  fetchGroup<Coin>() // Fails! There's evidence of Repository<Group<A>> but no evidence of `Repository<Coin>` available.
+}
+```
+
+We believe the encoding proposed above fits nicely with Kotlin's philosophy of extensions and will reduce the boilerplate compared to other langs that also support compile time dependency resolution such as Scala where this is done via implicits.
+
+## Language changes
+
+* Add `with` to require evidence of extensions in both function and class/object declarations.
+* Add `extension` to provide instance evidence for a given interface.
+
+Usage of these language changes are demonstrated by the previous and below examples:
+
+#### Class constraint
+
+```kotlin
+extension class GroupRepository<A>(with R: Repository<A>) : Repository<Group<A>> {
+  /* ... */
+}
+```
+
+#### Function constraint
+
+```kotlin
+fun <A> fetch(id: String, with R: Repository<A>): A = loadById(id) // function position using parameter "R"
+```
+
+#### Extension evidence using an Object
+
+```kotlin
+extension object UserRepository: Repository<User> {
+  override fun loadAll(): List<User> {
+    return listOf(User(25, "Bob"))
+  }
+
+  override fun loadById(id: Int): User? {
+    return if (id == 25) {
+      User(25, "Bob")
+    } else {
+      null
+    }
+  }
+}
+```
+
+#### Extension evidence using a Class
+
+```kotlin
+extension class UserRepository: Repository<User> {
+  override fun loadAll(): List<User> {
+    return listOf(User(25, "Bob"))
+  }
+
+  override fun loadById(id: Int): User? {
+    return if (id == 25) {
+      User(25, "Bob")
+    } else {
+      null
+    }
+  }
+}
+```
+
```

Comment:
Should we mention anything about compatibility with sealed, data, and inline classes here?

Reply:
That would require some exploration though. @truizlop do you know what's happening for those?

Reply:
@pakoito do you mean adding `extension` to a sealed / data / inline class, or providing an extension to one of those? I haven't explored the former case though; is there any use case they cover, that cannot be covered with the existing support?

Reply:
I mean whether they're compatible, yes.